rfcs/2023-01-06-layout.md

```diff
@@ -0,0 +1,460 @@
+
+In order to ease up building responsive UIs all the style prop values will also support an alternate **object based format** that accepts breakpoints specific values, keeping mobile-first approach (`@media(min-width)`). Sample usage:
+
+```tsx
+<Box
+  display="flex"
+  flexDirection={{ base: 'column', l: 'row' }}
+  alignItems="center"
+  marginTop="spacing.1"
+  marginBottom="spacing.2"
+  paddingX="spacing.1"
+  gap={{ base: 'spacing.1', m: 'spacing.2', l: 'spacing.3' }}
+>
+  <Button>Button 1</Button>
+  <Button>Button 2</Button>
+  <Button>Button 3</Button>
+</Box>
+```
+
+The keys are derived from **breakpoints tokens**. An extra `base` key can be added as a helper to define values for breakpoints below the ones provided, for example `{ base: 'column', l: 'row' }` implies the value is `column` for all breakpoints less than `l`. All wider viewports will inherit largest provided breakpoint value because of `min-width`.
+
+> **Note**
+>
+> Right now our breakpoints tokens are named as shorthands `s`, `m`, `l`, etc. We should keep the responsive properties same as breakpoints tokens. We can also consider making these device specific such as `smallMobile`, `largeMobile`, etc. to match the `useBreakpoints` hook. Tradeoff being, the current convention _(`xs`, `s`, `m`, `l`)_ sounds easier to compare _(is `largeMobile` larger than `smallTablet` or vice-versa)_. Changing these to `xsmall`, `small`, `medium` etc. can also work. Regardless, the responsive properties keys should match the breakpoints tokens.
+
+#### Layout based styling props
+
+All presentational components in Blade will support a subset of layout specific properties as top level props. The values will also be same as the ones supported in `Box`. Sample usage:
+
+```tsx
+<Box
+  display="flex"
+  flexDirection="column"
+  alignItems="center"
+  marginTop="spacing.1"
+  marginBottom="spacing.2"
+  paddingX="spacing.1"
+>
+  <Button>Button 1</Button>
+  <Button marginTop="spacing.1">Button 2</Button>
+  <Button marginTop="spacing.2">Button 3</Button>
+</Box>
+```
+
+Styling props will be the preferred choice for minor adjustments that can be done directly on the component without a need for creating extra `Box` wrapper _(which also keeps the DOM less nested and easier to read)_. Later on, we might want to expand the scope of these props to cover things beyond layout if need arises.
+
+As a tradeoff, styling props are pretty dangerous in terms of the flexibility they provide. They give power to customize any component to beyond recognition, for example you could customize the padding that comes on `Button`. In light of this capability we should support only a restricted set of properties to begin with _(which can't have adverse effect on look and feel of components)_.
+
+Styling props should be the preferred choice for minor adjustments only. You may note `padding*` is skipped since it can break existing padding of blade components. Supported restricted properties:
```

Comment:
<!-- Questions are appropriate if you have a potential concern but are not quite sure if it's relevant or not. Asking the author for clarification or investigation can lead to a quick resolution.  -->
**question:** ‏Can we elaborate a little on how padding on the Layout component breaks blade components? From my understanding, blade components will not support the styling props and only the layout components will 🤔 
An example/explanation would help.

Without padding, any non-blade component will not be able to use the Layout component at all.

Reply:
Added minor clarification in https://github.com/razorpay/blade/pull/950/commits/7633c183161d7d23854719d1259fff80b7f27617

`padding*` props will be available on the main layout component `Box` but intentionally skipped on presentational components such as `Button`, `Alert`, etc. where they may break component styling.

So non-blade components should still be able to use `Box` to build UIs.

Reply:
Works! 🙌🏽